### PR TITLE
[CPU] Improve computation of vector sizes

### PR DESCRIPTION
**Pull request by dcaballe**

### CODE REVIEW

**Comment by MacDue** [comment]
Hey! I know this is still WIP, but this looks like it'd help progress some of our scalable vectorization work. 

We noticed that the vectorizer would try to infer the vector shape from the IR before looking at the lowering config. For scalable tile sizes, it'd return oversized fixed vectors (we're not sure how it managed to find an upper bound) rather than scalable vectors. Switching to looking at the lowering config first (where we can represent the scalable sizes - see #15033) will make things much simpler. 

We also saw that `getCanonicalVectorShape()` appeared to look at the incorrect tiling levels and produced incorrect vector sizes, so switching to use `tilingConfig.getVectorTileSizes()` looks like it'll fix another one of our issues.

**Comment by dcaballe** [comment]
Thanks @MacDue... Yeah, we had a bunch of issues here. I also plan to introduce more changes in lowering config so maybe we should coordinate :).

(For some reason, I'm not able to add you as a reviewer but I would appreciate your feedback)

**Comment by MaheshRavishankar** [comment]
> Thanks @MacDue... Yeah, we had a bunch of issues here. I also plan to introduce more changes in lowering config so maybe we should coordinate :).
> 
> (For some reason, I'm not able to add you as a reviewer but I would appreciate your feedback)

It would be really useful for me to get an idea of what changes you are planning. Changes to lowering_config turn out to be disruptive downstream, so would be great to get an idea of the final state before starting to make changes. Also lowering config impacts all backends, so that needs to be accounted for.

**Comment by dcaballe** [comment]
> > Thanks @MacDue... Yeah, we had a bunch of issues here. I also plan to introduce more changes in lowering config so maybe we should coordinate :).
> > (For some reason, I'm not able to add you as a reviewer but I would appreciate your feedback)
> 
> It would be really useful for me to get an idea of what changes you are planning. Changes to lowering_config turn out to be disruptive downstream, so would be great to get an idea of the final state before starting to make changes. Also lowering config impacts all backends, so that needs to be accounted for.

Yes, for sure! I just brought this up because I saw that @MacDue is also introducing some changes to it and wanted to make sure we don't replicate work or go in opposite directions.

This PR should be ready to land. The RISC-V code size regressions are expected, as we were applying an ad-hoc extra level of tiling using an incorrect configuration. We'll find a way to fix that properly.

File: compiler/src/iree/compiler/Codegen/Common/GenericVectorization.cpp [hunk 1]
```diff
@@ -264,10 +194,14 @@ void GenericVectorizationPass::runOnOperation() {
   for (auto op : candidates) {
     SmallVector<int64_t> vectorSizes;
     if (auto linalgOp = dyn_cast<linalg::LinalgOp>(op)) {
-      // Do not vectorize the op if the vector size is greater than or eqaul
+      // Do not vectorize the op if the vector size is greater than or equal
```
Comment (banach-space) [inline compiler/src/iree/compiler/Codegen/Common/GenericVectorization.cpp:197]: [nit] Move just above L205

**Comment by banach-space** [comment]
LGTM, it's a rather straightforward clean-up of the logic for acquiring vector sizes before invoking the Linalg vectoriser. This has been a source of confusion for me for a while, so I'm really glad to see this being updated, thank you!

> It would be really useful for me to get an idea of what changes you are planning.

So, for a bit of context, we (Arm) are practically ready to enable scalable vectorisation in IREE. The only part that's missing is "teaching" IREE to be aware of scalability when e.g. creating/parsing the lowering config. That's implemented in 2 steps:
* https://github.com/openxla/iree/pull/15032
* https://github.com/openxla/iree/pull/15033

Once merged, we would be creating lowering configs as follows (note the additional `[` and `]` around `32`):
```
  #iree_codegen.lowering_config<tiling_levels
	= [[128, 128, 0], [1, [32], 0], [0, 0, 1], [0, 0, 0]]>
```
That's basically all we need for "scalability".  We'll need one more patch on top of that to wire things up. Here's a WIP version:
* https://github.com/MacDue/iree/commits/poc_scalable_vec

These changes shouldn't really affect other backends. And in general, we try to make sure that any changes specific to scalable vectors are only enabled for SVE (at leas for now). Of course, once we start experimenting and fine-tuning, we might come back with more ideas :)

**Review by banach-space — APPROVED** [review]

**Review by MaheshRavishankar — CHANGES_REQUESTED** [review]
Overall it looks fine. But there seem to be some real artifact size regressions. Do you know why?
Sorry for the delay, marking as request changes cause this might indicate a run-away vector size issue....

**Comment by dcaballe** [comment]
> Overall it looks fine. But there seem to be some real artifact size regressions. Do you know why? Sorry for the delay, marking as request changes cause this might indicate a run-away vector size issue....

No problem. Yep, it's what I mentioned above:

> This PR should be ready to land. The RISC-V code size regressions are expected, as we were applying an ad-hoc extra level of tiling using an incorrect configuration. We'll find a way to fix that properly.

There was an `if(isRISCV(...))` that was applying an extra level of tiling that was not present in the lowering config. Probably a gas stopper to land some changes that didn't work well on RISC-V. I prefer to deal with the regression and find the proper fix later.

**Review by MaheshRavishankar — APPROVED** [review]
Didnt notice these are RISC-V artifacts.

**Pull request merged**